### ISSUE TITLE
[Patterns] Inferred type of variable patterns in list pattern

### ISSUE THREAD

**Issue by scheglov** [dart-lang/language#2554]
Why do we infer `var a` as `Object? a`, not `int a`?
We said that we want `List<int>` with `<int>[var a]`.

```dart
  test_matchObject_withTypeArguments_variable_untyped() async {
    await assertNoErrorsInCode(r'''
void f(Object x) {
  switch (x) {
    case <int>[var a]:
      break;
  }
}
''');
    final node = findNode.switchPatternCase('case').pattern;
    assertResolvedNodeText(node, r'''
ListPattern
  typeArguments: TypeArgumentList
    leftBracket: <
    arguments
      NamedType
        name: SimpleIdentifier
          token: int
          staticElement: dart:core::@class::int
          staticType: null
        type: int
    rightBracket: >
  leftBracket: [
  elements
    VariablePattern
      keyword: var
      name: a
      declaredElement: hasImplicitType a@53
        type: Object?
  rightBracket: ]
  requiredType: List<int>
''');
  }
```

**Comment by lrhn** [comment]
Looks like a bug. I hink it should give `int`.

If there is no type argument on the list pattern, we should use the element type of the static type of the matched value as the matchee type used to do type inference on the element patterns.
If there is a type argument, we should probably use it instad.

@munificent

**Comment by lrhn** [comment]
Found this in the specification:

> ```dart
> var <int>[x] = ...
> ```
> 
> Here, we infer `int` for the inner `x` subpattern based on the type of the
> surrounding list pattern.

so that behavior is at least documented as being intended.

**Comment by munificent** [comment]
Fixed by #2569.
